src/System.Management.Automation/engine/Modules/GetModuleCommand.cs

```diff
@@ -482,14 +472,117 @@ private void GetLoadedModules(string[] names, IDictionary<string, ModuleSpecific
-                    modulesToWrite.Where(moduleInfo => ModuleMatch(moduleInfo, moduleSpecTable, PSEdition))
-                )
+            foreach (PSModuleInfo moduleInfo in FilterModulesForEditionAndSpecification(modulesToWrite, moduleSpecTable))
             {
                 WriteObject(moduleInfo);
             }
         }
+
+        /// <summary>
+        /// Filter an enumeration of PowerShell modules based on the required PowerShell edition
+        /// and the module specification constraints set for each module (if any).
+        /// </summary>
+        /// <param name="modules">The modules to filter through.</param>
+        /// <param name="moduleSpecificationTable">Module constraints, keyed by module name, to filter modules of that name by.</param>
+        /// <returns>All modules from the original input that meet both any module edition and module specification constraints provided.</returns>
+        private IEnumerable<PSModuleInfo> FilterModulesForEditionAndSpecification(
+            IEnumerable<PSModuleInfo> modules,
+            IDictionary<string, ModuleSpecification> moduleSpecificationTable)
+        {
+            if (!SkipEditionCheck)
+            {
+                modules = FilterModulesForIncompatibleOnEditionCheckedPath(modules);
+            }
+
+            if (!String.IsNullOrEmpty(PSEdition))
+            {
+                modules = FilterModulesForEdition(modules);
+            }
+
+            if (moduleSpecificationTable != null && moduleSpecificationTable.Count > 0)
+            {
+                modules = FilterModulesForSpecificationMatch(modules, moduleSpecificationTable);
+            }
+
+            return modules;
+        }
+
+        /// <summary>
+        /// Filter out all modules on the PowerShell-edition-checked path that are incompatible with
+        /// the current PowerShell edition from a given enumeration of modules.
+        /// </summary>
+        /// <param name="modules">The modules to filter incompatible examples from.</param>
+        /// <returns>All modules that are either not on the checked path or are compatible with the current PowerShell edition.</returns>
+        private IEnumerable<PSModuleInfo> FilterModulesForIncompatibleOnEditionCheckedPath(IEnumerable<PSModuleInfo> modules)
+        {
+            Dbg.Assert(!SkipEditionCheck, $"Caller to verify that {nameof(SkipEditionCheck)} is false");
+
+            string currentPSEdition = PSVersionInfo.PSEdition;
+            foreach (PSModuleInfo module in modules)
+            {
+                if (!module.IsLoadedFromCompatibilityCheckedPath)
+                {
+                    yield return module;
+                    continue;
+                }
+
+                IEnumerable<string> moduleCompatibleEditions = module.CompatiblePSEditions.Any() ? module.CompatiblePSEditions : DefaultCompatiblePSEditions;
+                if (moduleCompatibleEditions.Contains(currentPSEdition, StringComparer.OrdinalIgnoreCase))
```

Comment:
Going to change this over to the `Utils` method and optimise that one slightly. Want to let @daxian-dbw finish his review first (GitHub has had a nasty habit of forcing me to rewrite comments when new commits are added recently...)